### PR TITLE
Add usage tracking mechanism

### PR DESCRIPTION
**Pull request by guzman-raphael**
Depends on vathes/datajointio#69

Adds usage tracking mechanism to allow python packages to easily log user's package usage.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,17 +2,64 @@
...
+
+Since a reasonable solution could not be found, I introduced this feature that provides the mechanism and building blocks to have usage tracking data as granular as you need it.
+
+There are a few pre-requisites or assumptions:
+- User clients will have at least some periodic internet connection to upload usage logs to a centralized, remote usage-data aggregating host
+- 4 HTTP routes need to be implemented on your remote usage-data host
+  1. `GET` GUI-based authenticated route to register package installations with a user. You may use it to collect consent, have your user complete a survey, etc.
+  2. `POST` authenticated API route to accept the form submission of the above GUI route. An `installId` should be returned along with other details to ensure an 'open' connection.
+  3. `POST` authenticated API route that accepts user's event data and will store in an medium of your choice.
+  4. `POST` standard OAuth2.0 route that will allow refreshing `access_token`'s and `refresh_token`'s. PKCE flow currently implemented.
```
Comment (dimitri-yatsenko) [inline README.md:24]:
```suggestion
  4. `POST` standard OAuth2.0 route that will allow refreshing `access_token`'s and `refresh_token`'s. [PKCE flow](https://auth0.com/docs/flows/authorization-code-flow-with-proof-key-for-code-exchange-pkce) currently implemented.
```

File: README.md [hunk 2]
```diff
@@ -2,17 +2,64 @@
...
+The following features are currently supported.
+
+## Usage Tracking
+
+Have you ever wondered:
+
+- How many users are actually using my Python package?
+- How often are they using my Python package?
+- Which features/methods are most used and which ones are less interesting to the community?
+- Is there a better alternative to track and measure usage data than using anonymous download data available from PyPi's download logs in Google BigQuery? See [here](https://packaging.python.org/guides/analyzing-pypi-package-downloads/#background) for more details on this
```
Comment (dimitri-yatsenko) [inline README.md:14]:
```suggestion
- Is there a better alternative to track and measure usage data than using anonymous download data available from PyPi's download logs in Google BigQuery? See [Analyzing PyPI Package Downloads](https://packaging.python.org/guides/analyzing-pypi-package-downloads/#background) for more details on this
```

File: otumat/usage.py [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+from base64 import b64encode
+from time import sleep
+
+
+class UsageAgent:
+    """
+    Local agent which buffers package usage data locally and periodically sends logged data to
+    an OAuth2-compatible endpoint.
+    """
+    def __init__(self, author: str, data_directory: str, package_name: str, host: str = None,
```
Comment (dimitri-yatsenko) [inline otumat/usage.py:38]: Force keyword arguments for future proofing.
```suggestion
    def __init__(self, *, author: str, data_directory: str, package_name: str, host: str = None,
```

File: otumat/usage.py [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+        """
+        # determine period in seconds
+        period, unit = [int(e) if e.isdigit() else e
+                        for e in findall(r'([0-9]+)([a-z]+)', frequency)[0]]
+        if unit == 'm':
+            period *= 60
+        elif unit == 'h':
+            period *= 60 * 60
+        elif unit == 'd':
+            period *= 24 * 60 * 60
```
Comment (dimitri-yatsenko) [inline otumat/usage.py:433]: i would handle the `else` case here in case something other than `s` is used.

File: otumat/usage.py [hunk 5]
```diff
@@ -0,0 +1,528 @@
...
+    :param package_name: Name of package to identify the daemon
+    :type package_name: str
+    """
+    home_dir = getenv('USERPROFILE', getenv('HOME'))
+    if general_system() == 'Linux':
+        # disable startup by removing daemon command from profile script
+        # Bourne shell compatible
+        startup_file = Path(home_dir, '.profile')
+        if startup_file.exists():
+            lines = open(startup_file, 'r').readlines()
```
Comment (dimitri-yatsenko) [inline otumat/usage.py:512]: does this close the file? Should you use `with`?

File: otumat/usage.py [hunk 6]
```diff
@@ -0,0 +1,528 @@
...
+            with open(startup_file, 'w') as f:
+                [f.write(line) for line in lines if ('otumat' not in line and
+                                                     'upload' not in line and
+                                                     package_name not in line)]
```
Comment (dimitri-yatsenko) [inline otumat/usage.py:516]:
```suggestion
                for line in lines:
                    if not any(r in line for r in ('otumat', 'upload', package_name)):
                        f.write(line)
```

File: otumat/usage.py [hunk 7]
```diff
@@ -0,0 +1,528 @@
...
+    elif general_system() == 'Darwin':
+        # disable startup by removing property list file spec for the launch agent
+        startup_file = Path(home_dir, 'Library', 'LaunchAgents',
+                            f'{package_name}_usage.startup.plist')
+        if startup_file.exists():
```
Comment (dimitri-yatsenko) [inline otumat/usage.py:521]: don't check if exists first: unlink and catch error -- it's generally a better process for race conditions, etc.

File: otumat/usage.py [hunk 8]
```diff
@@ -0,0 +1,528 @@
...
+            startup_file.unlink()
+    elif general_system() == 'Windows':
+        # disable startup by removing the user's startup VBScript
+        startup_file = Path(home_dir, 'AppData', 'Roaming', 'Microsoft', 'Windows',
+                            'Start Menu', 'Programs', 'Startup', f'{package_name}_usage.vbs')
+        if startup_file.exists():
```
Comment (dimitri-yatsenko) [inline otumat/usage.py:527]: don't check for existence -- just catch exceptions.

File: otumat/usage.py [hunk 9]
```diff
@@ -0,0 +1,528 @@
...
+    """
+    home_dir = getenv('USERPROFILE', getenv('HOME'))
+    if general_system() == 'Linux':
+        # disable startup by removing daemon command from profile script
+        # Bourne shell compatible
+        startup_file = Path(home_dir, '.profile')
+        if startup_file.exists():
+            lines = open(startup_file, 'r').readlines()
+            with open(startup_file, 'w') as f:
+                [f.write(line) for line in lines if ('otumat' not in line and
```
Comment (dimitri-yatsenko) [inline otumat/usage.py:514]: can this remove extra entries that happen to have the package name in them?

**Review by dimitri-yatsenko — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (guzman-raphael) [reply README.md]: Addressed

File: README.md (hunk 2 above)
Comment (guzman-raphael) [reply README.md]: Addressed

File: otumat/usage.py (hunk 4 above)
Comment (guzman-raphael) [reply otumat/usage.py:458]: Addressed

File: otumat/usage.py (hunk 3 above)
Comment (guzman-raphael) [reply otumat/usage.py]: Good point, I've updated all methods to require kwargs now.

File: otumat/usage.py (hunk 5 above)
Comment (guzman-raphael) [reply otumat/usage.py]: It does not explicitly. I've now changed the pattern to utilize `Path(...).read_text()` instead which is cleaner and closes it appropriately w/o additional `with` statement.

File: otumat/usage.py (hunk 6 above)
Comment (guzman-raphael) [reply otumat/usage.py]: Thanks for the suggestion. I've updated it with a similar approach now.

File: otumat/usage.py (hunk 7 above)
Comment (guzman-raphael) [reply otumat/usage.py]: Ah yes, good point on race condition. I've updated it now to simply try to delete can catch the exception.

File: otumat/usage.py (hunk 8 above)
Comment (guzman-raphael) [reply otumat/usage.py]: Addressed

File: otumat/usage.py (hunk 9 above)
Comment (guzman-raphael) [reply otumat/usage.py]: I've updated this now. The prior implementation required certain terms to be present in the line. The new one now, requires specific flags, order, and spacing to be correct. Should be conservative enough.

**Comment by guzman-raphael** [comment]
@dimitri-yatsenko Thanks for the review. I've pushed updates addressing the following:
- Your feedback above
- Applied the styling updates for imports based on discussions in-person
- Changed the default flow to be more conservative than before i.e. assumes user does **not** want to participate in usage tracking unless they explicitly input `y` or `Y` in prompt. Non-interactive shells will assume no usage tracking.
- Included re-wording of console logs based feedback from 2021-04-30 DEMO
- Included `DISABLE_USAGE_TRACKING_PACKAGES` module property that allows certain modules to be explicitly excluded from usage tracking. This was from @ttngu207's feedback in the 2021-04-30 DEMO and will be useful when depending on packages that have usage tracking enabled.

**Review by dimitri-yatsenko — APPROVED** [review]

**Pull request merged**